eyecite/resolve.py

```diff
@@ -0,0 +1,180 @@
+from collections import defaultdict
+from typing import Callable, Dict, List, Optional, Tuple
+
+from eyecite.models import (
+    CitationBase,
+    FullCaseCitation,
+    IdCitation,
+    NonopinionCitation,
+    Resource,
+    ResourceType,
+    ShortCaseCitation,
+    SupraCitation,
+)
+from eyecite.utils import strip_punct
+
+
+def _resolve_fullcase_citation(citation: FullCaseCitation) -> Resource:
+    """
+    Resolve fullcase citations to resources directly.
+    """
+    return Resource(citation)
+
+
+def _resolve_shortcase_citation(
+    citation: ShortCaseCitation,
+    resolved_full_cites: Dict[ResourceType, FullCaseCitation],
+) -> Optional[ResourceType]:
+    """
+    Try to match shortcase citations by checking whether their reporter and
+    volume number matches those of any of the previously resolved full
+    citations. If there are multiple possible matches, try to refine by also
+    checking whether their antecedent_guess appears in either the defendant
+    or plaintiff field of any of the previously resolved full citations.
+    """
+    candidates: List[Tuple[ResourceType, FullCaseCitation]] = []
+    for resource, full_citation in resolved_full_cites.items():
+        if (
+            citation.reporter == full_citation.reporter
+            and citation.volume == full_citation.volume
+        ):
+            # Append both keys and values for further refinement below
+            candidates.append((resource, full_citation))
+
+    # Only accept if one candidate remains
+    if len(candidates) == 1:
+        return candidates[0][0]
+
+    # Otherwise, if there is an antecedent guess, try to refine further
+    elif citation.antecedent_guess:
+        refined_candidates: List[ResourceType] = []
+        ag: str = strip_punct(citation.antecedent_guess)
+
+        for resource, full_citation in candidates:
+            if full_citation.defendant and ag in full_citation.defendant:
+                refined_candidates.append(resource)
+            elif full_citation.plaintiff and ag in full_citation.plaintiff:
+                refined_candidates.append(resource)
+
+        # Only accept if one candidate remains
+        return refined_candidates[0] if len(refined_candidates) == 1 else None
```

Comment:
This feels like it'd be a candidate for factoring out.

Reply:
I see what you mean, but can you propose a name for the sub-function? I'm not quite sure what to call it.

Reply:
Something like, `filter_by_matching_antecedent`?